Doc/library/asyncio-task.rst

```diff
@@ -1064,6 +1066,36 @@ Task Object
       :meth:`cancel` and the wrapped coroutine propagated the
       :exc:`CancelledError` exception thrown into it.
 
+   .. method:: cancelling()
+
+      Return the number of cancellation requests to this Task, i.e.,
+      the number of calls to :meth:`cancel`.
+
+      Note that if this number is greater than zero but the Task is
+      still executing, :meth:`cancelled` will still return ``False``.
+      It's because this number can be lowered by calling :meth:`uncancel`,
+      which can lead to the task not being cancelled after all if the
+      cancellation requests go down to zero.
+
+   .. method:: uncancel()
+
+      Decrement the count of cancellation requests to this Task.
+
+      Returns the remaining number of cancellation requests.
+
+      This should be used by tasks that catch :exc:`CancelledError`
+      and wish to continue indefinitely until they are cancelled again::
+
+        async def resilient_task():
+            try:
+                await do_work()
+            except asyncio.CancelledError:
+                asyncio.current_task().uncancel()
+                await do_work()
```

Comment:
this example needs to link the cancellor to the cancellee, and show that callers of `task.uncancel()` also need to be responsible for scheduling a call to `.cancel()` and logging if they called it or not, and then always maintaining the `.cancel`/`.uncancel()` balance:

```python
async def make_request_with_timeout():
    loop = asyncio.get_running_loop()
    task = loop.current_task()
    i_called_cancel = False

    def on_timeout():
        nonlocal i_called_cancel
        i_called_cancel = True
        task.cancel()

    timeout_handle = loop.call_later(1, on_timeout)
    try:
        return await make_request()
    finally:
        timeout_handle.cancel()
        if (
            i_called_cancel
            and task.uncancel() == 0
            and sys.exc_info()[0] is asyncio.CancelledError
        ):
            raise TimeoutError
 ```

Reply:
@graingert I'm sure it's because I just am new to asyncio, but I don't understand what your comment (and example) has to do with a primary use-case of `uncancel()` from the perspective of Python users which is to simply ignore a cancellation and keep doing.

Why does the cancellor and cancellee need to be linked? The cancellee (child task) simply wants to proceed despite the parent trying to cancel. 

Why does the cancel/uncancel count need to be balanced? In fact, in [this comment](https://github.com/python/cpython/issues/90908#issuecomment-1093944756) Guido writes:

> A task that catches CancelledError is allowed to run undisturbed (ignoring further .cancel() calls and allowing any number of await calls!) until it either exits or calls .uncancel().


I've used trio before and am curious to see with 3.11's task groups how cancellation works. One positive change is that child tasks in a taskgroup can run despite other exceptions, and I would like some more real-world examples to fit my head around the suggested usage of this API. For example, I assumed previously that it worked like trio: when any child task in a taskgroup raised an exception then all others would be cancelled. And if all others were cancelled, they would primarily themselves raise CancelledError. So from this partial understanding I assumed that ExceptionGroups from a taskgroup would almost always only contain a single exception.

But now when I saw guidos comment, I more fully understand how ExceptionGroups from a taskgroup can realistically contain many things, and specifically I think this is because of the fact that TaskGroup child tasks can continue to run despite other cancellations.

Therefore, instead of an example with  `asyncio.timeout()` I would find it more useful to see an example of using taskgroups, timeouts, and handling cancellations altogether.

Maybe some pseudocode example of taskgroup that times out after 10 seconds and downloads all the images on a few web pages? Presumably if some image downloads failed, the other tasks would need to proceed.

Still I'm not sure whether in this instance the child tasks would _need_ to call `uncancel` (the docs don't yet explain why it's needed) or whether I can just handle the exception and proceed.

Reply:
> Why does the cancellor and cancellee need to be linked? The cancellee (child task) simply wants to proceed despite the parent trying to cancel.

consider two timeout context managers, outer and inner. inner expires first and then outer expires. If the current async function calls `uncancel()` it decrements the cancel counter early and corrupts the uncancel stack. Then later when the outer timeout expires the inner timeout context manager see the cancellation, decrements the uncancel() and incorrectly determines that it sent the cancel:

```python
import asyncio


async def amain():
    async with asyncio.timeout(2) as outer:
        try:
            async with asyncio.timeout(1) as inner:
                try:
                    await asyncio.sleep(2)
                except asyncio.CancelledError:
                    print("cancelled by inner")
                    asyncio.current_task().uncancel()  # corrupt the uncancel stack
                await asyncio.sleep(3)  # cancelled by outer
        except TimeoutError:
            raise AssertionError(
                "inner thought it cancelled the task but actually it was outer!"
            )


asyncio.run(amain())
```